WebEngine/Dialog.cpp

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "WebEngine/Dialog.h"
+
+#include <QHostAddress>
+#include <QMargins>
+#include <QWebChannel>
+#include <QWebEnginePage>
+#include <QWebEngineProfile>
+#include <QWebSocketServer>
+#include <memory>
+
+#include "JsonTransport.h"
+
+namespace WebEngine {
+
+Dialog::Dialog(QWebEngineProfile* profile, QWidget* parent) : QDialog(parent), page_(profile) {
+  web_socket_server_.reset(
+      // NOLINTNEXTLINE
+      new QWebSocketServer{"Web channel server", QWebSocketServer::NonSecureMode});
+  web_socket_server_->listen(QHostAddress::LocalHost, 12345);
```

Comment:
what if listen fails?

Reply:
Good point! Will think about something

Reply:
So, I made two changes:
- I changed the code to have the OS choose the listening port which should already make it hard to fail at that point. That was one thing I wanted to do anyway. Furthermore I added a getter for the WebSocketServer such that the port number can be requested by the consumer.
- I added a log message specifying the error.

This is obviously no hard error. A factory function returning a ErrorMessageOr would be good, but this doesn't work well with Qt-code because QObject-derived classes are neither copyable nor movable. I'm also not sure yet how this error should be handled best on the consumer side. It's probably not recoverable, so an error message box is probably the only solution. But I would prefer to do additional changes (if needed) when the consumer code is ready.

WDYT?

Reply:
sgtm..

question: do we need another solution for error handling, if ErrorMessageOr does not work here? Maybe have one that works with unique_ptr<_, DeleteLaterDeleter>?

Reply:
Good question. I was thinking a lot lately about that. In Qt almost all classes support an invalid state while we usually try to write stateless classes, meaning you can't construct them from invalid input data. And these two concepts don't compose well. I' not certain yet, but I think our code also needs to handle an invalid state when it's based on / interacts with Qt classes. But yeah, still thinking about it.